File: tests/zig/backends/wasm/graph_manifest_test.zig
//! Spec 01 — graph_manifest write / parse / emit / scan (single source of truth).
//! Embedded tests were intentionally not left in `src/wasm/graph_manifest.zig`
//! (that module is not a test root under the nested `wasm.*` package layout).

const std = @import("std");
const mathzig = @import("mathzig");
const gm = mathzig.wasm.graph_manifest;
const module_mod = mathzig.wasm.module;
const abi = mathzig.wasm.abi;

test "writeJson ↔ parse equality for fixture (T1)" {
    const inputs = [_]gm.Port{.{ .name = "x", .kind = .number }};
    const params = [_]gm.Port{.{ .name = "gain.k", .kind = .number, .default = 1.0 }};
    const outputs = [_]gm.GraphOutputPort{
        .{ .name = "u", .kind = .number, .result_tag = "number" },
        .{ .name = "v", .kind = .number, .result_tag = "number" },
    };
    const exports_ = [_]gm.ExportEntry{.{ .name = "tick", .params = 2 }};
    const m = gm.GraphManifest{
        .entry = "tick",
        .inputs = &inputs,
        .params = &params,
        .outputs = &outputs,
        .out_mode = .table,
        .exports = &exports_,
    };

    const json = try gm.toJsonAlloc(std.testing.allocator, m);
    defer std.testing.allocator.free(json);

    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const parsed = try gm.parseGraphManifestJson(arena.allocator(), json);

    try std.testing.expectEqual(@as(u32, 1), parsed.abi_version);
    try std.testing.expectEqualStrings("tick", parsed.entry);
    try std.testing.expectEqual(@as(usize, 1), parsed.inputs.len);
    try std.testing.expectEqualStrings("x", parsed.inputs[0].name);
    try std.testing.expectEqual(abi.WireKind.number, parsed.inputs[0].kind);
    try std.testing.expectEqual(@as(usize, 1), parsed.params.len);
    try std.testing.expectEqualStrings("gain.k", parsed.params[0].name);
    try std.testing.expect(parsed.params[0].default != null);
    try std.testing.expectApproxEqAbs(@as(f64, 1.0), parsed.params[0].default.?, 0);
    try std.testing.expectEqual(@as(usize, 2), parsed.outputs.len);
    try std.testing.expectEqualStrings("u", parsed.outputs[0].name);
    try std.testing.expectEqualStrings("v", parsed.outputs[1].name);
    try std.testing.expectEqual(gm.OutMode.table, parsed.out_mode);
    try std.testing.expectEqual(@as(usize, 1), parsed.exports.len);
    try std.testing.expectEqualStrings("tick", parsed.exports[0].name);
    try std.testing.expectEqual(@as(u32, 2), parsed.exports[0].params);
}

test "writeJson named_exports includes export field" {
    const outputs = [_]gm.GraphOutputPort{
        .{ .name = "u", .kind = .number, .export_name = "out_u" },
        .{ .name = "v", .kind = .matrix_ptr, .export_name = "out_v" },
    };
    const m = gm.GraphManifest{
        .outputs = &outputs,
        .out_mode = .named_exports,
        .exports = &[_]gm.ExportEntry{.{ .name = "out_u", .params = 0 }},
    };
    const json = try gm.toJsonAlloc(std.testing.allocator, m);
    defer std.testing.allocator.free(json);

    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const parsed = try gm.parseGraphManifestJson(arena.allocator(), json);
    try std.testing.expectEqual(gm.OutMode.named_exports, parsed.out_mode);
    try std.testing.expectEqualStrings("out_u", parsed.outputs[0].export_name.?);
    try std.testing.expectEqual(abi.WireKind.matrix_ptr, parsed.outputs[1].kind);
    try std.testing.expectEqualStrings("matrix", gm.kindToResultTag(parsed.outputs[1].kind));
}

test "emitCustomSection scan round-trip (T2)" {
    var mod = module_mod.WasmModule.init(std.testing.allocator);
    defer mod.deinit();

    const outputs = [_]gm.GraphOutputPort{
        .{ .name = "y", .kind = .number, .result_tag = "number" },
    };
    const m = gm.GraphManifest{
        .outputs = &outputs,
        .out_mode = .table,
        .exports = &[_]gm.ExportEntry{.{ .name = "tick", .params = 0 }},
    };
    try gm.emitCustomSection(&mod, std.testing.allocator, m);

    var out = std.Io.Writer.Allocating.init(std.testing.allocator);
    defer out.deinit();
    try mod.writeTo(&out.writer);

    const payload = gm.scanWasmGraphSection(out.written()) orelse {
        try std.testing.expect(false);
        return;
    };
    try std.testing.expectEqualStrings(gm.CUSTOM_SECTION_NAME, "mathzig:graph");
    try std.testing.expect(std.mem.indexOf(u8, payload, "\"name\":\"y\"") != null);
    try std.testing.expect(std.mem.indexOf(u8, payload, "\"out_mode\":\"table\"") != null);

    // Shared scanner: same payload via module.scanCustomSection
    const via_shared = module_mod.scanCustomSection(out.written(), gm.CUSTOM_SECTION_NAME);
    try std.testing.expect(via_shared != null);
    try std.testing.expectEqualStrings(payload, via_shared.?);
}

test "scanWasmGraphSection returns null when absent" {
    const bare = [_]u8{ 0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00 };
    try std.testing.expect(gm.scanWasmGraphSection(&bare) == null);
    try std.testing.expect(module_mod.scanCustomSection(&bare, "mathzig:node") == null);
}

test "validate rejects empty outputs; emit refuses empty (T5)" {
    try std.testing.expectError(error.EmptyOutputs, gm.validate(.{}));

    var mod = module_mod.WasmModule.init(std.testing.allocator);
    defer mod.deinit();
    try std.testing.expectError(error.EmptyOutputs, gm.emitCustomSection(&mod, std.testing.allocator, .{}));
}

test "parse rejects unknown kind and invalid out_mode (T5)" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();

    try std.testing.expectError(
        error.UnknownKind,
        gm.parseGraphManifestJson(a,
            \\{"outputs":[{"name":"y","kind":"not_a_kind"}]}
        ),
    );
    try std.testing.expectError(
        error.InvalidOutMode,
        gm.parseGraphManifestJson(a,
            \\{"outputs":[{"name":"y","kind":"number"}],"out_mode":"magic"}
        ),
    );
    try std.testing.expectError(
        error.EmptyOutputs,
        gm.parseGraphManifestJson(a,
            \\{"outputs":[]}
        ),
    );
}

test "graph_manifest reuses WireKind short names" {
    try std.testing.expectEqual(abi.WireKind.matrix_ptr, gm.parseKindName("matrix").?);
    try std.testing.expectEqualStrings("matrix", gm.kindShortName(.matrix_ptr));
    try std.testing.expectEqualStrings("matrix", gm.kindToResultTag(.matrix_ptr));
    try std.testing.expectEqual(gm.GRAPH_ABI_VERSION, @as(u32, 1));
}

File: tests/ts/graph_manifest.test.ts
/**
 * Spec 01 — `mathzig:graph` manifest read + validation path.
 * Section write lives in Zig (`src/wasm/graph_manifest.zig`).
 */
import { describe, expect, it } from "bun:test";
import {
  GRAPH_MANIFEST_SECTION,
  GraphManifestError,
  readGraphManifest,
  validateGraphManifest,
  type GraphManifest,
} from "../../src/ts/graph";

function encodeLeb128(n: number): number[] {
  const bytes: number[] = [];
  let value = n >>> 0;
  do {
    let b = value & 0x7f;
    value >>>= 7;
    if (value !== 0) b |= 0x80;
    bytes.push(b);
  } while (value !== 0);
  return bytes;
}

/** Minimal valid wasm module with a single custom section. */
function wasmWithCustomSection(name: string, content: string): Uint8Array {
  const enc = new TextEncoder();
  const nameBytes = enc.encode(name);
  const contentBytes = enc.encode(content);
  const payload = [...encodeLeb128(nameBytes.length), ...nameBytes, ...contentBytes];
  const section = [0, ...encodeLeb128(payload.length), ...payload];
  return new Uint8Array([0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00, ...section]);
}

const fixtureJson: GraphManifest = {
  abi: 1,
  entry: "tick",
  inputs: [{ name: "x", kind: "number" }],
  params: [{ name: "gain.k", kind: "number", default: 1.0 }],
  outputs: [
    { name: "u", kind: "number", result_tag: "number" },
    { name: "v", kind: "number", result_tag: "number" },
  ],
  out_mode: "table",
  exports: [{ name: "tick", params: 2 }],
};

describe("readGraphManifest", () => {
  it("returns null when mathzig:graph section is missing", async () => {
    const bare = new Uint8Array([0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00]);
    const module = await WebAssembly.compile(bare);
    expect(readGraphManifest(module)).toBeNull();
  });

  it("returns null for mathzig:node modules (wrong section)", async () => {
    const bytes = wasmWithCustomSection(
      "mathzig:node",
      JSON.stringify({
        abi: 1,
        name: "gain",
        inputs: [],
        params: [],
        output: { kind: "number", result_tag: "number" },
      }),
    );
    const module = await WebAssembly.compile(bytes);
    expect(readGraphManifest(module)).toBeNull();
  });

  it("parses mathzig:graph section written as fixture JSON", async () => {
    const bytes = wasmWithCustomSection(GRAPH_MANIFEST_SECTION, JSON.stringify(fixtureJson));
    const module = await WebAssembly.compile(bytes);
    const m = readGraphManifest(module);
    expect(m).not.toBeNull();
    expect(m!.abi).toBe(1);
    expect(m!.entry).toBe("tick");
    expect(m!.inputs).toEqual([{ name: "x", kind: "number" }]);
    expect(m!.params[0].name).toBe("gain.k");
    expect(m!.params[0].default).toBe(1.0);
    expect(m!.outputs).toHaveLength(2);
    expect(m!.outputs[0].name).toBe("u");
    expect(m!.outputs[1].result_tag).toBe("number");
    expect(m!.out_mode).toBe("table");
    expect(m!.exports).toEqual([{ name: "tick", params: 2 }]);
  });

  it("parses named_exports out_mode with export fields", async () => {
    const json: GraphManifest = {
      abi: 1,
      entry: "tick",
      inputs: [],
      params: [],
      outputs: [
        { name: "u", kind: "number", result_tag: "number", export: "out_u" },
        { name: "v", kind: "matrix", result_tag: "matrix", export: "out_v" },
      ],
      out_mode: "named_exports",
      exports: [
        { name: "out_u", params: 1 },
        { name: "out_v", params: 1 },
      ],
    };
    const bytes = wasmWithCustomSection(GRAPH_MANIFEST_SECTION, JSON.stringify(json));
    const module = await WebAssembly.compile(bytes);
    const m = readGraphManifest(module);
    expect(m!.out_mode).toBe("named_exports");
    expect(m!.outputs[0].export).toBe("out_u");
    expect(m!.outputs[1].kind).toBe("matrix");
  });

  it("throws on empty outputs in section payload (T5)", async () => {
    const bad = { ...fixtureJson, outputs: [] as GraphManifest["outputs"] };
    const bytes = wasmWithCustomSection(GRAPH_MANIFEST_SECTION, JSON.stringify(bad));
    const module = await WebAssembly.compile(bytes);
    expect(() => readGraphManifest(module)).toThrow(GraphManifestError);
  });
});

describe("validateGraphManifest (T5)", () => {
  it("accepts fixture", () => {
    const m = validateGraphManifest(fixtureJson);
    expect(m.outputs).toHaveLength(2);
    expect(m.out_mode).toBe("table");
  });

  it("rejects empty outputs", () => {
    expect(() => validateGraphManifest({ ...fixtureJson, outputs: [] })).toThrow(
      /outputs must be non-empty/,
    );
  });

  it("rejects unknown kind as GraphManifestError", () => {
    let caught: unknown;
    try {
      validateGraphManifest({
        ...fixtureJson,
        outputs: [{ name: "y", kind: "not_a_kind" }],
      });
    } catch (e) {
      caught = e;
    }
    expect(caught).toBeInstanceOf(GraphManifestError);
    expect(String((caught as Error).message)).toMatch(/Unknown port kind/);
  });

  it("rejects invalid out_mode", () => {
    expect(() =>
      validateGraphManifest({
        ...fixtureJson,
        out_mode: "magic" as GraphManifest["out_mode"],
      }),
    ).toThrow(/out_mode/);
  });
});
